internal/wasm/jit/asm/amd64/impl.go

```diff
@@ -0,0 +1,401 @@
+		}
+	case operandTypesConstToMemory:
+		if n.dstMemIndex != asm.NilRegister {
+			ret = fmt.Sprintf("%s 0x%x, [%s + 0x%x + %s*0x%x]", instName, n.srcConst,
+				registerName(n.dstReg), n.dstConst, registerName(n.dstMemIndex), n.dstMemScale)
+		} else {
+			ret = fmt.Sprintf("%s 0x%x, [%s + 0x%x]", instName, n.srcConst, registerName(n.dstReg), n.dstConst)
+		}
+	case operandTypesConstToRegister:
+		ret = fmt.Sprintf("%s 0x%x, %s", instName, n.srcConst, registerName(n.dstReg))
+	}
+	return
+}
+
+type operandType byte
+
+const (
+	operandTypeNone operandType = iota
+	operandTypeRegister
+	operandTypeMemory
+	operandTypeConst
+	operandTypeBranch
+)
+
+func (o operandType) String() (ret string) {
+	switch o {
+	case operandTypeNone:
+		ret = "none"
+	case operandTypeRegister:
+		ret = "register"
+	case operandTypeMemory:
+		ret = "memory"
+	case operandTypeConst:
+		ret = "const"
+	case operandTypeBranch:
+		ret = "branch"
+	}
+	return
+}
+
+type operandTypes struct{ src, dst operandType }
+
+var (
+	operandTypesNoneToNone         = operandTypes{operandTypeNone, operandTypeNone}
+	operandTypesNoneToRegister     = operandTypes{operandTypeNone, operandTypeRegister}
+	operandTypesNoneToMemory       = operandTypes{operandTypeNone, operandTypeMemory}
+	operandTypesNoneToBranch       = operandTypes{operandTypeNone, operandTypeBranch}
+	operandTypesRegisterToNone     = operandTypes{operandTypeRegister, operandTypeNone}
+	operandTypesRegisterToRegister = operandTypes{operandTypeRegister, operandTypeRegister}
+	operandTypesRegisterToMemory   = operandTypes{operandTypeRegister, operandTypeMemory}
+	operandTypesRegisterToConst    = operandTypes{operandTypeRegister, operandTypeConst}
+	operandTypesMemoryToRegister   = operandTypes{operandTypeMemory, operandTypeRegister}
+	operandTypesMemoryToConst      = operandTypes{operandTypeMemory, operandTypeConst}
+	operandTypesConstToRegister    = operandTypes{operandTypeConst, operandTypeRegister}
+	operandTypesConstToMemory      = operandTypes{operandTypeConst, operandTypeMemory}
+)
+
+func (o operandTypes) String() string {
+	return fmt.Sprintf("from:%s,to:%s", o.src, o.dst)
+}
+
+// assemblerImpl implements Assembler.
+type assemblerImpl struct {
+	asm.BaseAssemblerImpl
+
+	root, current *nodeImpl
+}
+
+var _ Assembler = &assemblerImpl{}
+
+// newNode creates a new Node and appends it into the linked list.
+func (a *assemblerImpl) newNode(instruction asm.Instruction, types operandTypes) *nodeImpl {
+	n := &nodeImpl{
+		instruction: instruction,
+		prev:        nil,
+		next:        nil,
+		types:       types,
+	}
+	a.addNode(n)
+	return n
+}
+
+// addNode appends the new node into the linked list.
+func (a *assemblerImpl) addNode(node *nodeImpl) {
+	if a.root == nil {
+		a.root = node
+		a.current = node
+	} else {
+		parent := a.current
+		parent.next = node
+		a.current = node
+	}
+}
+
+func (a *assemblerImpl) encodeNode(w io.Writer, n *nodeImpl) (err error) {
```

Comment:
as node functions sometimes traverse, maybe mention that this only encodes the current node, so that you don't have to click in to figure that out. unless it will traverse :D